File: src/control_robot/src/control.cpp
//* Standard libraries
#include <cstddef>
#include <cstdint>
#include <opencv2/core/mat.hpp>
#include <opencv2/core/types.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <string>
#include <utility>
#include <vector>

//* ROS2
#include "cv_bridge/cv_bridge.h"
#include "rclcpp/rclcpp.hpp"
#include <rclcpp/logging.hpp>
#include <rclcpp/publisher.hpp>
#include <rclcpp/subscription.hpp>

#include "geometry_msgs/msg/twist.hpp"
#include <custom_interfaces/msg/prediction.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <sensor_msgs/msg/image.hpp>

//* Boost libraries
#include <boost/circular_buffer.hpp>
#include <boost/math/statistics/linear_regression.hpp>

class BasicControl : public rclcpp::Node {

private:
  static constexpr const char *NODE_NAME = "control_basic_node";
  // static constexpr const char *COMMAND_VELOCITY_TOPIC = "/j100_0395/cmd_vel";
  static constexpr const char *COMMAND_VELOCITY_TOPIC = "cmd_vel";
  static constexpr const char *LANE_DETECTION_IMG_TOPIC =
      "lane_detection_output";
  static constexpr const char *PREDICTION_TOPIC_SUB = "prediction_video";
  static constexpr const int STANDARD_QOS = 10;

  static constexpr const size_t AMOUNT_BOTTOMMOST_POINTS = 7;
  static constexpr const int IMAGE_WIDTH = 640 - 1;
  static constexpr const int IMAGE_HEIGHT = 360 - 1;
  static constexpr const int IMAGE_CENTER = IMAGE_WIDTH / 2;
  static constexpr const float MAX_ANG_VEL = 1.0f;
  static constexpr const float MAX_LIN_VEL = 0.5f;
  static constexpr const float K_VALUE = 0.011111111f;
  static constexpr const int BUFFER_CAPACITY =
      30; //* Will hold only the 30 most relevant points

  static constexpr const int X = 0;
  static constexpr const int Y = 1;

  using Twist = geometry_msgs::msg::Twist;
  using Prediction = custom_interfaces::msg::Prediction;
  using Point = geometry_msgs::msg::Point;
  using Image = sensor_msgs::msg::Image;
  using Lane = std::vector<std::vector<int>>;

  rclcpp::Publisher<Twist>::SharedPtr pub_cmd_vel_;
  rclcpp::Publisher<Image>::SharedPtr pub_result_img;
  rclcpp::Subscription<Prediction>::SharedPtr sub_prediction_;

  boost::circular_buffer<float> left_lane_buffer_;
  boost::circular_buffer<float> right_lane_buffer_;

  /**
   * @brief Receives a vector of points, and returns the same values in a vector
   * of ints
   *
   * @param point_vector
   * @return Lane
   */
  std::pair<Lane, Lane>
  obtain_lanes(const std::vector<Point> &points_left_lane,
               const std::vector<Point> &points_right_lane) {
    Lane left_lane;
    left_lane.reserve(points_left_lane.size());
    for (const auto &point : points_left_lane) {
      //* If point.x is a negative value, it means that is out of bounds
      if (point.x >= 0)
        left_lane.push_back(
            {static_cast<int>(point.x), static_cast<int>(point.y)});
      else
        left_lane.push_back({0, static_cast<int>(point.y)});
    }

    Lane right_lane;
    right_lane.reserve(points_right_lane.size());
    for (const auto &point : points_right_lane) {
      //* If point.x is a negative value, it means that is out of bounds
      if (point.x >= 0)
        right_lane.push_back(
            {static_cast<int>(point.x), static_cast<int>(point.y)});
      else
        right_lane.push_back({IMAGE_WIDTH, static_cast<int>(point.y)});
    }

    //* Sort array based on y-value, from greatest to lowest
    auto descending_sort = [](const std::vector<int> &a,
                              const std::vector<int> &b) {
      return a[1] > b[1];
    };
    std::sort(left_lane.begin(), left_lane.end(), descending_sort);
    std::sort(right_lane.begin(), right_lane.end(), descending_sort);

    return {left_lane, right_lane};
  }

  /**
   * @brief Returns the middle row between two pairs of lanes
   *
   * @param left_lane
   * @param right_lane
   * @return Lane
   */
  Lane middle_row_between(const Lane &left_lane, const Lane &right_lane) {
    size_t indx_l = 0;
    size_t indx_r = 0;
    Lane middle_row;

    while (indx_l < left_lane.size() && indx_r < right_lane.size()) {
      //* If the y-coordinates match, get middle row
      if (left_lane[indx_l][Y] == right_lane[indx_r][Y]) {
        int middle_x = (left_lane[indx_l][X] + right_lane[indx_r][X]) / 2;
        middle_row.push_back({middle_x, left_lane[indx_l][Y]});
        indx_l++;
        indx_r++;
      }
      //* If the y-coordinates do not match, then continue until they do
      else {
        if (left_lane[indx_l][Y] > right_lane[indx_r][Y])
          indx_l++; //* Advance l because list is in descending order, and
                    // greater means need to advance to reduce
        else
          indx_r++;
      }
    }

    return middle_row;
  }

  void publish_and_visualize(const Image &img, const Lane lane1,
                             const Lane lane2, const Lane middle_lane,
                             int average, int angle_degrees, float dx, float dy,
                             float lin_vel, float ang_vel) {
    cv_bridge::CvImagePtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvCopy(img, sensor_msgs::image_encodings::BGR8);
    } catch (cv_bridge::Exception &e) {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
      return;
    }

    int THICKNESS = 3;
    cv::polylines(cv_ptr->image, lane1, true, cv::Scalar(0, 255, 0), THICKNESS,
                  cv::LINE_8);
    cv::polylines(cv_ptr->image, lane2, true, cv::Scalar(0, 255, 0), THICKNESS,
                  cv::LINE_8);
    cv::polylines(cv_ptr->image, middle_lane, true, cv::Scalar(0, 0, 255),
                  THICKNESS, cv::LINE_8);

    //* Draw a line from top center of image, to the bottom of the image where
    // the lane_center was calculated. Show it in red
    size_t size = middle_lane.size();
    cv::line(cv_ptr->image, cv::Point(IMAGE_CENTER, IMAGE_HEIGHT),
             cv::Point(average, middle_lane[size - 1][Y]),
             cv::Scalar(255, 0, 0), THICKNESS);

    // Text box with debug info
    std::string text = "A:" + std::to_string(angle_degrees) +
                       " dx:" + std::to_string((int)dx) +
                       " dy:" + std::to_string((int)dy);
    int text_x = IMAGE_CENTER - 80;
    int text_y = 30;
    cv::Size text_size =
        cv::getTextSize(text, cv::FONT_HERSHEY_SIMPLEX, 0.5, 1, nullptr);
    cv::rectangle(cv_ptr->image,
                  cv::Point(text_x - 5, text_y - text_size.height - 5),
                  cv::Point(text_x + text_size.width + 5, text_y + 5),
                  cv::Scalar(255, 255, 255), -1);
    cv::putText(cv_ptr->image, text, cv::Point(text_x, text_y),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 0));

    // Second text box with velocity info
    std::string vel_text = "lin:" + std::to_string(lin_vel).substr(0, 4) +
                           " ang:" + std::to_string(ang_vel).substr(0, 4);
    int vel_y = text_y + 25;
    cv::Size vel_text_size =
        cv::getTextSize(vel_text, cv::FONT_HERSHEY_SIMPLEX, 0.5, 1, nullptr);
    cv::rectangle(cv_ptr->image,
                  cv::Point(text_x - 5, vel_y - vel_text_size.height - 5),
                  cv::Point(text_x + vel_text_size.width + 5, vel_y + 5),
                  cv::Scalar(255, 255, 255), -1);
    cv::putText(cv_ptr->image, vel_text, cv::Point(text_x, vel_y),
                cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 0));

    pub_result_img->publish(*cv_ptr->toImageMsg());
  }

  /**
   * @brief Obtain the prediction message, and perform the control defined
   * 
   * @param prediction_msg 
   */
  void prediction_subscriber(const Prediction::ConstSharedPtr &prediction_msg) {
    const std::vector<Point> &predicted_left_lane = prediction_msg->left_lane;
    const std::vector<Point> &predicted_right_lane = prediction_msg->right_lane;
    const Image &image = prediction_msg->frame;

    const auto [left_lane, right_lane] =
        obtain_lanes(predicted_left_lane, predicted_right_lane);

    Lane middle_lane = middle_row_between(left_lane, right_lane);
    if (middle_lane.size() == 0)
      return;

    middle_lane.resize(
        std::min(middle_lane.size(),
                 AMOUNT_BOTTOMMOST_POINTS)); //* Truncate the array to the first
                                             // AMOUNT_BOTTOMMOST_POINTS

    //* Obtain average of the first 7 points
    int average = 0;
    int num_points = middle_lane.size();
    for (const auto &point : middle_lane)
      average += point[X];
    average /= num_points;

    float dx = average - IMAGE_CENTER; // Horizontal difference
    float dy = IMAGE_HEIGHT -
               middle_lane[num_points - 1][Y]; // Vertical difference (negative
                                               // since y increases downward)

    float angle_radians = std::atan2(dy, dx);
    float angle_degrees = angle_radians * 180.0 / M_PI;
    float diff = angle_degrees - 90; //* Bring angle to [-90,90]

    //* Control calculation
    float lin_vel = (num_points / 7.0) * MAX_LIN_VEL;
    // float w = (diff * diff) / (90 * 90);
    float w = K_VALUE * diff;
    float ang_vel = MAX_ANG_VEL * w;

    if (diff < 0)
      ang_vel *= -1;

    lin_vel = lin_vel * (1 - w);
    publish_vel(lin_vel, ang_vel);
    publish_and_visualize(image, left_lane, right_lane, middle_lane, average,
                          diff, dx, dy, lin_vel, ang_vel);
  }

  /**
   * @brief Function solely to recieve linear and angular velocity, and publish ti
   * 
   * @param linear_x 
   * @param ang_z 
   */
  void publish_vel(float linear_x, float ang_z) {
    Twist msg;
    msg.linear.set__x(linear_x);
    msg.angular.set__z(ang_z);

    pub_cmd_vel_->publish(msg);
  }

public:
  explicit BasicControl() : Node(NODE_NAME) {
    left_lane_buffer_ = boost::circular_buffer<float>(BUFFER_CAPACITY);
    right_lane_buffer_ = boost::circular_buffer<float>(BUFFER_CAPACITY);

    //* Initialize publisher and subscriber
    pub_cmd_vel_ =
        this->create_publisher<Twist>(COMMAND_VELOCITY_TOPIC, STANDARD_QOS);
    pub_result_img =
        this->create_publisher<Image>(LANE_DETECTION_IMG_TOPIC, STANDARD_QOS);
    sub_prediction_ = this->create_subscription<Prediction>(
        PREDICTION_TOPIC_SUB, STANDARD_QOS,
        [this](const Prediction::ConstSharedPtr &prediction) {
          this->prediction_subscriber(prediction);
        });
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);

  auto node = std::make_shared<BasicControl>();
  rclcpp::spin(node);

  rclcpp::shutdown();

  return 0;
}

File: src/pycontrol/pycontrol/control.py
import rclpy
import numpy as np
from rclpy.node import Node
from custom_interfaces.msg import Prediction
from cv_bridge import CvBridge
from geometry_msgs.msg import Twist, Point
from sensor_msgs.msg import Image
import cv2

class PyControl(Node):
    def __init__(self):
        super().__init__('pycontroller')

        self.pred = Prediction()
        self.vel = Twist()
        self.image = Image()
        self.left_line = []
        self.right_line = []
        self.left_lane = []
        self.right_lane = []
    
        self.current_time = 0.0
        self.last_time = 0.0

        self.pred_subscribe = self.create_subscription(Prediction, 'prediction_video', self.pred_cb, 10)

        # self.im_pub = self.create_publisher(Image, 'prediction_video', 10)
        # self.vel_pub = self.create_publisher(Twist, 'cmd_vel', 10)
        self.vel_pub = self.create_publisher(Twist, '/j100_0395/cmd_vel', 10)
        self.pub_result_img = self.create_publisher(Image, '/lane_detection_output', 10)

        timer_period = 0.05
        self.timer = self.create_timer(timer_period, self.timer_cb)

    def pred_cb(self, msg):
        self.pred = msg
        self.left_lane = self.pred.left_lane
        self.right_lane = self.pred.right_lane
        self.image = self.pred.frame
        self.get_logger().info(f'entered pred_cb')
        
    def publish_and_visualize(self, img, lane1, lane2, middle_lane, average, 
                     angle_degrees, dx, dy, lin_vel, ang_vel, no_pred = False):
        """
        Visualize lane detection results and publish the annotated image.
        
        Args:
            img: ROS Image message
            lane1: First lane polyline points (numpy array or list)
            lane2: Second lane polyline points (numpy array or list)
            middle_lane: Middle lane polyline points (numpy array or list)
            average: Average x-coordinate for lane center
            angle_degrees: Steering angle in degrees
            dx: Delta x displacement
            dy: Delta y displacement
            lin_vel: Linear velocity
            ang_vel: Angular velocity
            no_pred: If True, only show velocity info and no prediction message
        """
        bridge = CvBridge()
        
        try:
            cv_image = bridge.imgmsg_to_cv2(img, desired_encoding='bgr8')
        except:
            return
        
        # Variables to plot information
        IMAGE_CENTER = self.image.width // 2
        IMAGE_HEIGHT = self.image.height
        angle_degrees = int((angle_degrees * 180) / np.pi)
        
        # Show this view when no prediction was made
        if no_pred:
            # No prediction case - only show velocity and warning message
            
            # Warning message at top left
            no_pred_text = "No Prediction"
            cv2.putText(cv_image, no_pred_text, (0, 20),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 2)
            
            # Velocity info
            vel_text = f"lin:{lin_vel:.3f} ang:{ang_vel:.3f}"
            text_x = IMAGE_CENTER - 80
            vel_y = 50
            
            (vel_text_width, vel_text_height), _ = cv2.getTextSize(
                vel_text, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
            
            cv2.rectangle(cv_image,
                        (text_x - 5, vel_y - vel_text_height - 5),
                        (text_x + vel_text_width + 5, vel_y + 5),
                        (255, 255, 255), -1)
            cv2.putText(cv_image, vel_text, (text_x, vel_y),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0))
            
        else:
            # Normal prediction case - show all visualizations
            
            THICKNESS = 3
            Y = 1  # Index for y-coordinate in point array
            
            # Draw lane polylines
            cv2.polylines(cv_image, [np.array(lane1, dtype=np.int32)], 
                        False, (0, 255, 0), THICKNESS, cv2.LINE_8) # prediction 1
            cv2.polylines(cv_image, [np.array(lane2, dtype=np.int32)], 
                        False, (0, 255, 0), THICKNESS, cv2.LINE_8) # prediction 2
            cv2.polylines(cv_image, [np.array(middle_lane, dtype=np.int32)], 
                        False, (0, 0, 255), THICKNESS, cv2.LINE_8) # middle lane
            
            
            
            # Blue line that goes from center to lane center
            size = len(middle_lane)
            self.get_logger().info(f'Middle lane: {middle_lane}')
            cv2.line(cv_image, 
                    (IMAGE_CENTER, IMAGE_HEIGHT),
                    (int(average), int(middle_lane[size - 1, Y])),
                    (255, 0, 0), THICKNESS)
            
            # Text box with debug info
            text = f"Error:{angle_degrees} dx:{int(dx)} dy:{int(dy)}"
            text_x = IMAGE_CENTER - 80
            text_y = 30
            
            (text_width, text_height), _ = cv2.getTextSize(
                text, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
            
            cv2.rectangle(cv_image,
                        (text_x - 5, text_y - text_height - 5),
                        (text_x + text_width + 5, text_y + 5),
                        (255, 255, 255), -1)
            cv2.putText(cv_image, text, (text_x, text_y),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0))
            
            # Second text box with velocity info
            vel_text = f"lin:{lin_vel:.3f} ang:{ang_vel:.3f}"
            vel_y = text_y + 25
            
            (vel_text_width, vel_text_height), _ = cv2.getTextSize(
                vel_text, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
            
            cv2.rectangle(cv_image,
                        (text_x - 5, vel_y - vel_text_height - 5),
                        (text_x + vel_text_width + 5, vel_y + 5),
                        (255, 255, 255), -1)
            cv2.putText(cv_image, vel_text, (text_x, vel_y),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0))
        
        # Publish the annotated image
        result_msg = bridge.cv2_to_imgmsg(cv_image, encoding='bgr8')
        self.pub_result_img.publish(result_msg)

    def timer_cb(self):
        self.convert_line()
        self.calculate_middle()

        self.get_logger().info(f'{len(self.middle_row)}')
        middle_line = np.array(self.middle_row)
        middle_line = middle_line[:min(len(middle_line), 7)]
        
        self.get_logger().debug(f'len: {len(middle_line)}')

        if len(middle_line) == 0:
            self.current_time = self.get_clock().now().nanoseconds / 1e9
            if self.current_time - self.last_time > 10.0:
                self.vel.linear.x = 0.0
                self.vel.angular.z = 0.0
                self.get_logger().info(f'entered middle_line_0')

                self.vel_pub.publish(self.vel)
                # debug
                self.publish_and_visualize(self.image, [], [], [], 0, 0, 0, 0, self.vel.linear.x, self.vel.angular.z, True)
            else:
                self.vel_pub.publish(self.vel)
                # debug
                self.publish_and_visualize(self.image, middle_line, [], [], 0, 0, 0, 0, self.vel.linear.x, self.vel.angular.z, True)
        
        else:
            average = middle_line[:, 0].mean() if len(middle_line) > 0 else 0
            dx = average - self.image.width // 2
            dy = self.image.height - middle_line[:, 1].mean() if len(middle_line) > 0 else 0
            
            angle = np.arctan2(dy, dx)
            w_error = angle - np.pi / 2 # Between -pi/2 and pi/2

            if abs(w_error) > np.deg2rad(5):
                Kp = 0.3
                Kw = 0.6

                self.vel.linear.x = (len(middle_line) / 7.0) * Kp 
                self.vel.angular.z = w_error * Kw

                self.vel_pub.publish(self.vel)
                self.get_logger().info(f'Linear Velocity: {self.vel.linear.x:.2f} Angular Velocity: {self.vel.angular.z:.2f}')
                self.last_time = self.get_clock().now().nanoseconds / 1e9
                # debug
                self.publish_and_visualize(self.image, self.left_line, self.right_line, middle_line, average, w_error, dx, dy, self.vel.linear.x, self.vel.angular.z)

            else:
                Kp = 0.5
                Kw = 0.3

                self.vel.linear.x = (len(middle_line) / 7.0) * Kp 
                self.vel.angular.z = w_error * Kw

                self.vel_pub.publish(self.vel)
                self.get_logger().info(f'Linear Velocity: {self.vel.linear.x:.2f} Angular Velocity: {self.vel.angular.z:.2f}')
                self.last_time = self.get_clock().now().nanoseconds / 1e9
                # debug
                self.publish_and_visualize(self.image, self.left_line, self.right_line, middle_line, average, w_error, dx, dy, self.vel.linear.x, self.vel.angular.z)
                
    def convert_line(self):

        self.left_line = []
        self.right_line = []

        for point in self.left_lane:
            if point.x > 0 :
                self.left_line.append((point.x, point.y))
            else:
                self.left_line.append((0, point.y))

        for point in self.right_lane:
            if point.x > 0 :
                self.right_line.append((point.x, point.y))
            else:
                self.right_line.append((0, point.y))

        self.right_line.sort(reverse=True)
        self.left_line.sort(reverse=True)

    def calculate_middle(self):
        indx_l = 0
        indx_r = 0
        self.middle_row = []
        
        # if len(self.left_line) > 3:
        #     self.get_logger().info(f"Left line: [({self.left_line[0][0]},{self.left_line[0][1]}),({self.left_line[1][0]},{self.left_line[1][1]}),({self.left_line[2][0]},{self.left_line[2][1]})...({self.left_line[-1][0]},{self.left_line[-1][1]})]")
        # else:
        #     self.get_logger().info(f"Left line: {self.left_line}")

        # Print right line
        # if len(self.right_line) > 3:
        #     self.get_logger().info(f"Right line: [({self.right_line[0][0]},{self.right_line[0][1]}),({self.right_line[1][0]},{self.right_line[1][1]}),({self.right_line[2][0]},{self.right_line[2][1]})...({self.right_line[-1][0]},{self.right_line[-1][1]})]")
        # else:
        #     self.get_logger().info(f"Right line: {self.right_line}")

        
        while indx_l < len(self.left_line) and indx_r < len(self.right_line):
            # If the y-coordinates match, get middle row
            if abs(self.left_line[indx_l][1] - self.right_line[indx_r][1]) < 20.0:
                middle_x = (self.left_line[indx_l][0] + self.right_line[indx_r][0]) // 2  # X coordinate
                self.middle_row.append((middle_x, self.left_line[indx_l][1]))
                indx_l += 1
                indx_r += 1
            # If the y-coordinates do not match, then continue until they do
            else:
                if self.left_line[indx_l][1] > self.right_line[indx_r][1]:
                    indx_l += 1  # Advance l because list is in descending order
                else:
                    indx_r += 1


def main(args=None):
    rclpy.init(args=args)
    pycon = PyControl()
    rclpy.spin(pycon)
    pycon.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()